docs/source/schemas/dynamic_columns.rst

```diff
@@ -0,0 +1,140 @@
+===============
+Dynamic Columns
+===============
+
+.. note::
+
+    This document discusses AtlasDB's dynamic columns. This is a separate concept from Cassandra's dynamic columns;
+    in fact, AtlasDB's schemas in Cassandra always use two Cassandra dynamic columns, independent of whether
+    we are storing a fixed-columns or dynamic columns table! Throughout this document, where "dynamic columns" is
+    used unqualified, we refer to AtlasDB dynamic columns.
+
+Overview
+--------
+
+AtlasDB offers support for *dynamic columns* - in tables where these are enabled, rows that can have an arbitrary number
+of columns, each with a (typed) column name. A table with dynamic columns may be reasoned about as a
+``Map<RowKey, SortedMap<Tuple<Col1, Col2... ColN>, Value>>``. For example, a simple schema using dynamic columns may be
+defined as follows:
+
+.. code:: java
+
+    schema.addTableDefinition("todo", new TableDefinition() {{
+                rowName();
+                    rowComponent("person", ValueType.STRING);
+                dynamicColumns();
+                    columnComponent("taskSize", ValueType.VAR_LONG);
+                    columnComponent("monetaryCost", ValueType.VAR_LONG);
+                    value(ValueType.STRING);
+            }});
+
+Note that dynamic column components must be primitive ValueTypes which support partitioning and ordering.
+Also, key-value pairs in an individual row will be returned in *lexicographically* sorted order of the key.
+
+Dynamic Columns are useful for avoiding KVS-level range scans, especially in key-value services where range scans
+are expensive (like Cassandra - its caching optimisations are rendered ineffectual for range scans). For example,
+for the schema defined above:
+
+1. "Find John's smallest and cheapest todo" can be answered very efficiently.
+   Elements are stored in sorted order, so the very first key-value pair retrieved from the database is actually the
+   correct answer to this query.
+2. "Find all of Tom's todos with size up to M, limited to N results" is easy, as this can be
+   processed with a natural stopping point, as in query 1, and we are iterating in the right order for this query.
+   This can be achieved using the getRowsColumnRange API.
+3. "Find all of Tom's todos with size of 10 to 15, limit to N results" can be performed via a range
+   scan on the dynamic column *values*. This is also readily supported.
```

Comment:
do our APIs actually support scanning based on the value? maybe i'm getting confused on the terminology; it seems to be that the `taskSize` is part of the column *key*, and the *value* is a string

Reply:
Ah, this is a terminology thing. I generally have used "key" for the row-key, and "column values" for the column keys. We should standardise on this (row-key, column key, value).

Also, for full disambiguation I'll add something like "find all of X's descriptions starting with Foo". That needs scanning through the entire row, since there's no range scan on values.